File: collector/modules/instagram.py
"""LaclauGPT-native Instagram parser.

This parser follows the original LaclauGPT-TikTok-Scraper style: identify the
visited view and captured request, collect native media objects, then normalise
one object at a time. It supports several current Instagram API/GraphQL shapes
without mirroring another collector's parser structure.

Zeeschuimer is an architectural inspiration for browser/API-response capture,
but this module is an independent LaclauGPT implementation and is not a port of
Zeeschuimer source code.

Author: Tomi Toivio / LaclauGPT
License: CC0 1.0 Universal
"""
from __future__ import annotations

import json
import re
from datetime import datetime, timezone
from typing import Any
from urllib.parse import urlparse

from .common import as_string, first_value, unique

MODULE_NAME = "Instagram (posts & reels)"
DOMAIN = "instagram.com"

_HASHTAG_RE = re.compile(r"#([\wÀ-ÖØ-öø-ÿ]+)", re.UNICODE)
_RESERVED_PATHS = {
    "p", "reel", "reels", "explore", "stories", "direct", "accounts",
    "popular", "about", "legal", "developer", "web",
}


def _host(url: str) -> str:
    try:
        return urlparse(url).netloc.lower().removeprefix("www.")
    except (TypeError, ValueError):
        return ""


def _page_context(url: str) -> tuple[str, str | None, str | None]:
    """Return (view, profile_handle, requested_shortcode)."""
    try:
        path = [part for part in urlparse(url).path.split("/") if part]
    except (TypeError, ValueError):
        return "unknown", None, None
    if not path:
        return "frontpage", None, None
    head = path[0].lower()
    if head == "explore":
        return "explore", None, None
    if head == "p" and len(path) > 1:
        return "single_post", None, path[1]
    if head in {"reel", "reels"} and len(path) > 1:
        return "single_reel", None, path[1]
    if head not in _RESERVED_PATHS:
        if len(path) > 1 and path[1].lower() == "reels":
            return "user_reels", head, None
        if len(path) > 1 and path[1].lower() == "tagged":
            return "user_tagged", head, None
        return "user_posts", head, None
    return head, None, None


def _caption_text(item: dict) -> str:
    caption = item.get("caption")
    if isinstance(caption, str):
        return caption
    if isinstance(caption, dict):
        return as_string(caption.get("text"))
    if isinstance(caption, list) and caption and isinstance(caption[0], dict):
        return as_string(caption[0].get("text"))
    try:
        return as_string(item["edge_media_to_caption"]["edges"][0]["node"]["text"])
    except (KeyError, IndexError, TypeError):
        return ""


def _looks_like_media(item: Any) -> bool:
    if not isinstance(item, dict):
        return False
    identity = first_value(item.get("pk"), item.get("id"), item.get("code"), item.get("shortcode"))
    if not identity:
        return False
    typename = as_string(item.get("__typename"))
    return bool(
        item.get("media_type")
        or item.get("code")
        or item.get("shortcode")
        or item.get("image_versions2")
        or item.get("video_versions")
        or item.get("display_url")
        or item.get("display_uri")
        or item.get("edge_media_to_caption")
        or "Polaris" in typename
    )


def _collect_media(root: Any) -> list[dict]:
    """Collect post/reel objects while treating carousel children as media assets."""
    found: list[dict] = []
    seen_objects: set[int] = set()

    def walk(node: Any) -> None:
        if isinstance(node, list):
            for child in node:
                walk(child)
            return
        if not isinstance(node, dict):
            return
        marker = id(node)
        if marker in seen_objects:
            return
        seen_objects.add(marker)

        if _looks_like_media(node):
            found.append(node)
            return

        for key, value in node.items():
            if key in {
                "user", "owner", "caption", "audio", "music_metadata", "image_versions2",
                "video_versions", "carousel_media", "edge_sidecar_to_children",
            }:
                continue
            if isinstance(value, (dict, list)):
                walk(value)

    walk(root)
    return found


def _parse_html(text: str) -> list[dict]:
    payloads: list[dict] = []
    for match in re.finditer(
        r"<script[^>]*type=[\"']application/json[\"'][^>]*>([\s\S]*?)</script>",
        text,
        re.IGNORECASE,
    ):
        try:
            value = json.loads(match.group(1))
        except (json.JSONDecodeError, TypeError):
            continue
        if isinstance(value, dict):
            payloads.append(value)
    return [item for payload in payloads for item in _collect_media(payload)]


def _is_ad(item: dict) -> bool:
    link = item.get("link")
    return bool(
        item.get("product_type") == "ad"
        or item.get("ad_action") is not None
        or (isinstance(link, str) and link.startswith("https://www.facebook.com/ads/"))
    )


def _author_name(item: dict) -> str:
    user = item.get("user") if isinstance(item.get("user"), dict) else {}
    owner = item.get("owner") if isinstance(item.get("owner"), dict) else {}
    return as_string(first_value(user.get("username"), owner.get("username"), item.get("owner_username")))


def _decorate(item: dict, view: str, embedded: bool) -> dict:
    copied = dict(item)
    copied["_laclaugpt_partial"] = not bool(
        _caption_text(item)
        and (
            item.get("video_versions")
            or item.get("image_versions2")
            or item.get("display_url")
            or item.get("display_uri")
        )
    )
    copied["_laclaugpt_instagram_view"] = view
    copied["_laclaugpt_embedded_json"] = embedded
    return copied


def _deduplicate(items: list[dict]) -> list[dict]:
    by_id: dict[str, dict] = {}
    for item in items:
        item_id = as_string(first_value(item.get("pk"), item.get("id"), item.get("code"), item.get("shortcode")))
        if not item_id:
            continue
        previous = by_id.get(item_id)
        if previous is None or (previous.get("_laclaugpt_partial") and not item.get("_laclaugpt_partial")):
            by_id[item_id] = item
    return list(by_id.values())


def capture(response: Any, source_platform_url: str, source_url: str) -> list[dict]:
    """Extract native Instagram post/reel objects from one captured response."""
    if _host(source_platform_url) != DOMAIN:
        return []

    view, visited_handle, requested_shortcode = _page_context(source_platform_url)
    request_url = source_url or ""
    if re.search(r"logging_client_events|lightspeed_web_request_for_igd|injected_story_units", request_url):
        return []

    embedded = False
    if isinstance(response, dict):
        items = _collect_media(response)
    elif isinstance(response, str):
        body = response.strip()
        if not body:
            return []
        if body.startswith("for (;;);"):
            body = body[len("for (;;);"):]
        try:
            payload = json.loads(body)
        except json.JSONDecodeError:
            embedded = True
            items = _parse_html(body)
        else:
            items = _collect_media(payload)
    else:
        return []

    collected: list[dict] = []
    for item in items:
        if _is_ad(item):
            continue
        shortcode = as_string(first_value(item.get("code"), item.get("shortcode")))
        if requested_shortcode and shortcode != requested_shortcode:
            continue
        author = _author_name(item).lower()
        if visited_handle and author and author != visited_handle.lower():
            continue
        collected.append(_decorate(item, view, embedded))
    return _deduplicate(collected)


def overwrite_partial(incoming: dict, existing: dict | None) -> bool:
    """Allow a full payload to replace an earlier partial payload, never reverse."""
    if not existing:
        return False
    return bool(existing.get("_laclaugpt_partial")) and not bool(
        incoming.get("_laclaugpt_partial")
    )


def _media_type(item: dict) -> str:
    typename = as_string(item.get("__typename"))
    if item.get("media_type") == 8 or item.get("carousel_media") or item.get("edge_sidecar_to_children"):
        return "carousel"
    if item.get("media_type") == 2 or item.get("is_video") or "VideoMedia" in typename:
        return "video"
    return "photo"


def _image_urls(item: dict) -> list[str]:
    urls: list[str] = []
    for key in ("display_url", "display_uri", "display_src"):
        if item.get(key):
            urls.append(as_string(item[key]))
    image_versions = item.get("image_versions2")
    if isinstance(image_versions, dict):
        for candidate in image_versions.get("candidates") or []:
            if isinstance(candidate, dict) and candidate.get("url"):
                urls.append(as_string(candidate["url"]))

    children = item.get("carousel_media") or []
    if isinstance(children, list):
        for child in children:
            if isinstance(child, dict):
                urls.extend(_image_urls(child))
    return unique(urls)


def _video_urls(item: dict) -> list[str]:
    urls: list[str] = []
    if item.get("video_url"):
        urls.append(as_string(item["video_url"]))
    for version in item.get("video_versions") or []:
        if isinstance(version, dict) and version.get("url"):
            urls.append(as_string(version["url"]))
    children = item.get("carousel_media") or []
    if isinstance(children, list):
        for child in children:
            if isinstance(child, dict):
                urls.extend(_video_urls(child))
    return unique(urls)


def _timestamp(item: dict) -> tuple[str, int]:
    value = first_value(item.get("taken_at"), item.get("taken_at_timestamp"), item.get("created_time"))
    try:
        unix_ts = int(str(value))
    except (TypeError, ValueError):
        return "", 0
    if unix_ts <= 0:
        return "", unix_ts
    try:
        return datetime.fromtimestamp(unix_ts, timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"), unix_ts
    except (OverflowError, OSError, ValueError):
        return "", 0


def _location(item: dict) -> tuple[str, str, str, str]:
    loc = item.get("location") if isinstance(item.get("location"), dict) else {}
    name = as_string(loc.get("name"))
    city = as_string(first_value(loc.get("city"), loc.get("city_name")))
    location_id = as_string(first_value(loc.get("pk"), loc.get("id")))
    lat = first_value(loc.get("lat"), loc.get("latitude"))
    lng = first_value(loc.get("lng"), loc.get("longitude"))
    latlong = f"{lat},{lng}" if lat is not None and lng is not None else ""
    return name, latlong, city, location_id


def map_item(item: dict, metadata: dict | None = None) -> dict:
    """Map one native Instagram object to the stable collector record shape."""
    metadata = metadata or {}
    user = item.get("user") if isinstance(item.get("user"), dict) else {}
    owner = item.get("owner") if isinstance(item.get("owner"), dict) else {}
    author = {**owner, **{key: value for key, value in user.items() if value is not None}}

    shortcode = as_string(first_value(item.get("code"), item.get("shortcode")))
    native_id = as_string(first_value(item.get("pk"), item.get("id"), shortcode))
    record_id = shortcode or native_id
    media_type = _media_type(item)
    caption = _caption_text(item)
    timestamp, unix_ts = _timestamp(item)
    image_urls = _image_urls(item)
    video_urls = _video_urls(item)
    location_name, location_latlong, location_city, location_id = _location(item)

    if shortcode:
        route = "reel" if media_type == "video" else "p"
        permalink = f"https://www.instagram.com/{route}/{shortcode}"
    else:
        permalink = as_string(first_value(item.get("permalink"), item.get("link")))

    hashtags = unique(match.group(1) for match in _HASHTAG_RE.finditer(caption))
    likes = first_value(
        item.get("like_count"),
        (item.get("edge_media_preview_like") or {}).get("count"),
        (item.get("edge_liked_by") or {}).get("count"),
    )
    comments = first_value(
        item.get("comment_count"),
        (item.get("edge_media_to_comment") or {}).get("count"),
        (item.get("edge_media_to_parent_comment") or {}).get("count"),
    )
    plays = first_value(item.get("play_count"), item.get("view_count"), item.get("video_view_count"))

    return {
        "collected_from_url": metadata.get("source_platform_url", ""),
        "id": record_id,
        "native_id": native_id,
        "thread_id": record_id,
        "parent_id": "",
        "author": as_string(first_value(author.get("username"), item.get("owner_username"))),
        "author_full": as_string(first_value(author.get("full_name"), author.get("name"))),
        "author_id": as_string(first_value(author.get("pk"), author.get("id"))),
        "author_avatar": as_string(
            first_value(author.get("profile_pic_url"), author.get("profile_pic_url_hd"))
        ),
        "verified": author.get("is_verified", ""),
        "body": caption,
        "timestamp": timestamp,
        "unix_timestamp": unix_ts,
        "url": permalink,
        "media_type": media_type,
        "media_urls": ",".join(video_urls),
        "image_urls": ",".join(image_urls),
        "hashtags": ",".join(hashtags),
        "num_likes": likes if likes is not None else -1,
        "num_comments": comments if comments is not None else -1,
        "like_count": likes if likes is not None else -1,
        "comment_count": comments if comments is not None else -1,
        "play_count": plays if plays is not None else -1,
        "location_name": location_name,
        "location_latlong": location_latlong,
        "location_city": location_city,
        "location_id": location_id,
        "product_type": as_string(item.get("product_type")),
        "partial": bool(item.get("_laclaugpt_partial")),
        "instagram_view": as_string(item.get("_laclaugpt_instagram_view")),
    }

File: collector/browser/modules/instagram.js
/**
 * LaclauGPT Instagram parser.
 *
 * The structure deliberately follows the original LaclauGPT scraper style:
 * identify the request, route it to a small parser, normalise one platform
 * object at a time, and emit LaclauGPT records. Instagram has changed its web
 * payloads repeatedly, so this module accepts several current GraphQL/API
 * shapes without copying another collector's parser architecture.
 *
 * Author: Tomi Toivio / LaclauGPT
 * License: CC0 1.0 Universal
 */

function asString(value) {
  return value === null || value === undefined ? "" : String(value);
}

function firstValue(...values) {
  return values.find(value => value !== null && value !== undefined && value !== "");
}

function unique(values) {
  return [...new Set(values.filter(Boolean))];
}

function toIsoTimestamp(value) {
  const numeric = Number(value);
  if (!Number.isFinite(numeric) || numeric <= 0) return null;
  try {
    return new Date(numeric * 1000).toISOString();
  } catch {
    return null;
  }
}

function resolveUrls(sourcePlatformUrl, sourceUrl) {
  const candidates = [sourcePlatformUrl, sourceUrl].filter(value => typeof value === "string" && value);
  const requestUrl = candidates.find(url => /instagram\.com\/(graphql|api|ajax)\b/i.test(url)) ?? sourcePlatformUrl ?? sourceUrl ?? "";
  const pageUrl = candidates.find(url => url !== requestUrl && /instagram\.com/i.test(url)) ?? sourceUrl ?? sourcePlatformUrl ?? "";
  return { requestUrl, pageUrl };
}

function profileHandle(pageUrl) {
  try {
    const path = new URL(pageUrl).pathname.split("/").filter(Boolean);
    if (!path.length) return null;
    const reserved = new Set([
      "p", "reel", "reels", "explore", "stories", "direct", "accounts",
      "popular", "about", "legal", "developer", "web",
    ]);
    if (reserved.has(path[0].toLowerCase())) return null;
    return path[0].toLowerCase();
  } catch {
    return null;
  }
}

function captionText(item) {
  const caption = item?.caption;
  if (typeof caption === "string") return caption;
  if (caption && typeof caption.text === "string") return caption.text;
  if (Array.isArray(caption) && typeof caption[0]?.text === "string") return caption[0].text;

  const edgeCaption = item?.edge_media_to_caption?.edges?.[0]?.node?.text;
  if (typeof edgeCaption === "string") return edgeCaption;
  return "";
}

function imageCandidates(item) {
  const urls = [];
  if (item?.display_url) urls.push(item.display_url);
  if (item?.display_src) urls.push(item.display_src);
  for (const candidate of item?.image_versions2?.candidates ?? []) {
    if (candidate?.url) urls.push(candidate.url);
  }
  return urls;
}

function videoCandidates(item) {
  const urls = [];
  if (item?.video_url) urls.push(item.video_url);
  for (const version of item?.video_versions ?? []) {
    if (version?.url) urls.push(version.url);
  }
  return urls;
}

function collectMedia(item) {
  const urls = [...videoCandidates(item), ...imageCandidates(item)];
  const children = item?.carousel_media ?? item?.edge_sidecar_to_children?.edges?.map(edge => edge?.node) ?? [];
  for (const child of children) {
    if (!child) continue;
    urls.push(...videoCandidates(child));
    urls.push(...imageCandidates(child));
  }
  return unique(urls);
}

function mediaType(item) {
  if (item?.media_type === 8 || item?.carousel_media || item?.edge_sidecar_to_children) return "carousel";
  if (item?.media_type === 2 || item?.is_video || /VideoMedia/i.test(item?.__typename ?? "")) return "video";
  return "image";
}

function parseInstagramItem(item, via, pageUrl) {
  if (!item || typeof item !== "object") return null;

  const user = item.user ?? item.owner ?? {};
  const shortcode = asString(firstValue(item.code, item.shortcode));
  const nativeId = asString(firstValue(item.pk, item.id, shortcode));
  if (!nativeId) return null;

  const username = firstValue(user.username, item.owner_username);
  const caption = captionText(item);
  const permalink = shortcode
    ? `https://www.instagram.com/${mediaType(item) === "video" ? "reel" : "p"}/${shortcode}/`
    : firstValue(item.permalink, item.link);

  return {
    post_id: nativeId,
    author: username ?? null,
    author_display: firstValue(user.full_name, user.name, username) ?? null,
    timestamp: toIsoTimestamp(firstValue(item.taken_at, item.taken_at_timestamp, item.created_time)),
    url: permalink ?? null,
    text: caption,
    hashtags: unique(caption.match(/#[\p{L}\p{N}_]+/gu) ?? []),
    engagement: {
      likes: firstValue(item.like_count, item.edge_media_preview_like?.count, item.edge_liked_by?.count),
      comments: firstValue(item.comment_count, item.edge_media_to_comment?.count, item.edge_media_to_parent_comment?.count),
      views: firstValue(item.view_count, item.play_count, item.video_view_count),
      reshares: firstValue(item.reshare_count, item.repost_count),
    },
    media_urls: collectMedia(item),
    media_type: mediaType(item),
    instagram: {
      shortcode: shortcode || null,
      native_id: nativeId,
      product_type: item.product_type ?? null,
      user_id: asString(firstValue(user.pk, user.id)) || null,
      user_verified: user.is_verified ?? null,
      user_followers: firstValue(user.follower_count, user.edge_followed_by?.count),
      accessibility_caption: item.accessibility_caption ?? null,
      page_url: pageUrl || null,
    },
    via,
  };
}

function looksLikeMedia(item) {
  if (!item || typeof item !== "object" || Array.isArray(item)) return false;
  const hasIdentity = Boolean(firstValue(item.pk, item.id, item.code, item.shortcode));
  const hasMediaSignal = Boolean(
    item.media_type || item.code || item.shortcode || item.image_versions2 ||
    item.video_versions || item.display_url || item.edge_media_to_caption ||
    /Polaris(?:Post|VideoMedia|ImageMedia)/i.test(item.__typename ?? "")
  );
  return hasIdentity && hasMediaSignal;
}

function collectInstagramItems(root) {
  const items = [];
  const seenObjects = new Set();

  function walk(node) {
    if (!node || typeof node !== "object") return;
    if (seenObjects.has(node)) return;
    seenObjects.add(node);

    if (Array.isArray(node)) {
      for (const child of node) walk(child);
      return;
    }

    if (looksLikeMedia(node)) {
      items.push(node);
      // Carousel children are media assets belonging to this post, not posts.
      return;
    }

    const preferredKeys = [
      "data", "items", "medias", "feed_items", "fill_items", "edges",
      "repost_grid_items", "two_by_two_item", "node", "media",
      "xdt_api__v1__feed__user_timeline_graphql_connection",
      "xdt_api__v1__feed__timeline__connection",
    ];

    let followedPreferredKey = false;
    for (const key of preferredKeys) {
      if (node[key] !== undefined) {
        followedPreferredKey = true;
        walk(node[key]);
      }
    }

    // GraphQL envelopes often introduce generated wrapper names. Only if none
    // of the known containers were present, descend one level through objects.
    if (!followedPreferredKey) {
      for (const [key, value] of Object.entries(node)) {
        if (["user", "owner", "caption", "audio", "music_metadata"].includes(key)) continue;
        if (value && typeof value === "object") walk(value);
      }
    }
  }

  walk(root);
  return items;
}

function dataGrabber(data, requestUrl, pageUrl) {
  if (!data || typeof data !== "object") return [];

  // Avoid obvious telemetry/background payloads. The old LaclauGPT scraper
  // routed only known content endpoints; this is the equivalent guard here.
  if (/logging_client_events|lightspeed_web_request_for_igd|injected_story_units/i.test(requestUrl)) {
    return [];
  }

  const visitedHandle = profileHandle(pageUrl);
  const records = [];
  for (const item of collectInstagramItems(data)) {
    const record = parseInstagramItem(item, /graphql/i.test(requestUrl) ? "graphql" : "api", pageUrl);
    if (!record) continue;

    // LaclauGPT contamination guard: when explicitly collecting a profile,
    // ignore preloaded/suggested posts from other accounts.
    if (visitedHandle && record.author && record.author.toLowerCase() !== visitedHandle) {
      continue;
    }
    records.push(record);
  }
  return records;
}

function parseEmbeddedHtml(html, pageUrl) {
  const records = [];
  const scripts = html.matchAll(/<script[^>]*type=["']application\/json["'][^>]*>([\s\S]*?)<\/script>/gi);
  for (const match of scripts) {
    try {
      records.push(...dataGrabber(JSON.parse(match[1]), "embedded-html", pageUrl));
    } catch {
      // Ignore unrelated or malformed application/json blocks.
    }
  }
  return records;
}

function deduplicate(records) {
  const byId = new Map();
  for (const record of records) {
    if (!record?.post_id) continue;
    const previous = byId.get(record.post_id);
    if (!previous || (record.media_urls?.length ?? 0) > (previous.media_urls?.length ?? 0)) {
      byId.set(record.post_id, record);
    }
  }
  return [...byId.values()];
}

export async function parse(response, sourcePlatformUrl, sourceUrl) {
  const { requestUrl, pageUrl } = resolveUrls(sourcePlatformUrl, sourceUrl);
  const records = [];

  if (response && typeof response === "object") {
    return deduplicate(dataGrabber(response, requestUrl, pageUrl));
  }
  if (typeof response !== "string" || !response.trim()) return [];

  let body = response.trim();
  if (body.startsWith("for (;;);")) body = body.slice("for (;;);".length);

  try {
    records.push(...dataGrabber(JSON.parse(body), requestUrl, pageUrl));
  } catch {
    records.push(...parseEmbeddedHtml(body, pageUrl));
  }

  return deduplicate(records);
}
